component/loki/process/internal/stages/metric.go

```diff
@@ -0,0 +1,385 @@
+package stages
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component/loki/process/internal/metric"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+)
+
+// Metric types.
+const (
+	MetricTypeCounter   = "counter"
+	MetricTypeGauge     = "gauge"
+	MetricTypeHistogram = "histogram"
+
+	defaultMetricsPrefix = "loki_process_custom_"
+)
+
+// Configuration errors.
+var (
+	ErrEmptyMetricsStageConfig = errors.New("empty metric stage configuration")
+	ErrMetricsStageInvalidType = errors.New("invalid metric type: must be one of 'counter', 'gauge', or 'histogram'")
+	ErrInvalidIdleDur          = errors.New("max_idle_duration could not be parsed as a time.Duration")
+	ErrSubSecIdleDur           = errors.New("max_idle_duration less than 1s not allowed")
+)
+
+// MetricConfig is a single metrics configuration.
+// TODO(@tpaschalis) Rework once River squashing is implemented.
+type MetricConfig struct {
+	Counter   *metric.CounterConfig   `river:"counter,block,optional"`
+	Gauge     *metric.GaugeConfig     `river:"gauge,block,optional"`
+	Histogram *metric.HistogramConfig `river:"histogram,block,optional"`
+}
+
+// UnmarshalRiver implements the unmarshaller
+func (mc *MetricConfig) UnmarshalRiver(f func(v interface{}) error) error {
+	*mc = MetricConfig{}
+	type metricCfg MetricConfig
+	err := f((*metricCfg)(mc))
+	if err != nil {
+		return err
+	}
+
+	nonEmpty := 0
+	v := reflect.Indirect(reflect.ValueOf(mc))
```

Comment:
Would it be more efficient to check the three fields for nil?

Reply:
If there's no objection, I'd leave this as-is for now for consistency with the way we handle the StageConfig struct. Once struct squashing is implemented in River, they should both be amended to use it (so we can say `stage.match` and `metric.gauge` respectively)